File: [K]graph/[K]graph-bfs/542-01-matrix.py
from collections import deque
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        rows, cols = len(mat), len(mat[0])
        res = [[0 for _ in range(cols)] for _ in range(rows)]

        visited = set()
        queue = deque([])
        for r in range(rows):
            for c in range(cols):
                if mat[r][c] == 0:
                    queue.append((r, c, 0))
                    visited.add((r, c))
        
        while queue:
            r, c, step = queue.popleft()
            res[r][c] = step
            for next_r, next_c in [(r+1, c), (r-1, c), (r, c+1), (r, c-1)]:
                if next_r not in range(rows) or next_c not in range(cols) or (next_r, next_c) in visited:
                    continue
                queue.append((next_r, next_c, step + 1))
                visited.add((next_r, next_c))
        return res
    
# time O(RC)
# space O(RC)
# using graph and bfs with multiple sources

File: [K]graph/[K]graph-bfs/542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int rows = mat.length;
        int cols = mat[0].length;
        int[][] res = new int[rows][cols];

        ArrayDeque<int[]> queue = new ArrayDeque<>();
        HashSet<Integer> visited = new HashSet<>();

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (mat[r][c] == 0) {
                    queue.offer(new int[]{r, c, 0});
                    visited.add(r * cols + c);
                }
            }
        }
        
        while (!queue.isEmpty()) {
            int[] rcs = queue.poll();
            int r = rcs[0];
            int c = rcs[1];
            int s = rcs[2];
            res[r][c] = s;
            for (int[] nextrc: new int[][]{{r + 1, c}, {r - 1, c}, {r, c + 1}, {r, c - 1}}) {
                int nextr = nextrc[0];
                int nextc = nextrc[1];
                if (0 <= nextr && nextr < rows && 0 <= nextc && nextc < cols && !visited.contains(nextr * cols + nextc)) {
                    queue.offer(new int[]{nextr, nextc, s + 1});
                    visited.add(nextr * cols + nextc);
                }
            }
        }
        return res;
    }
}

// time O(RC)
// space O(RC)
// using graph and bfs with multiple sources